### ISSUE TITLE
[Usage]: Logits processor with v1

### ISSUE THREAD

**Issue by Famondir** [vllm-project/vllm#21672]
### Your current environment

Hello everyone,
I can't find an example on how to use the new logits processor class for vllm v1. I want to build something that probably look like this with the old framework:

It should work similar to guided decoding.

```
        def process_token(token_ids, logits):
            for i, token_id in enumerate(token_ids):
                # print(f"Token ID: {token_id}, Logit: {logits[i]}")
                if token_id not in [
                    self.tokenizer.encode("no", add_special_tokens=False)[0],
                    self.tokenizer.encode("oth", add_special_tokens=False)[0],
                    self.tokenizer.encode("Ak", add_special_tokens=False)[0],
                    self.tokenizer.encode("Pas", add_special_tokens=False)[0],
                    self.tokenizer.encode("Gu", add_special_tokens=False)[0]
                ]:
                    logits[i] = -np.inf
            return logits
```

I try to achieve this because I'm searching for the different behaviour in the predictions of vllm with guided decoding vs transformers with prefix_allowed_tokens_fn. I describe the case in the [gemma3-27b-it repositoory](https://huggingface.co/google/gemma-3-27b-it/discussions/70#6881317792a7d591385c618f).

I hope you can help me with an example or an idea why the classification with Guided Decoding in vllm performs so much wore that the transformers approach.

Edit: The bad performance with vllm 0.9.2 was caused by a bug in the on board transformers version as described [in this thread](https://github.com/vllm-project/vllm/issues/21034#issuecomment-3078791043).

Best regards
Simon

```text
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version                : Could not collect
CMake version                : version 4.0.3
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.7.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.11 (main, Jun  4 2025, 08:56:18) [GCC 11.4.0] (64-bit runtime)
Python platform              : Linux-6.8.0-60-generic-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   : LAZY
GPU models and configuration : GPU 0: NVIDIA H100 PCIe
Nvidia driver version        : 570.133.20
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               56
On-line CPU(s) list:                  0-55
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7453 28-Core Processor
CPU family:                           25
Model:                                1
Thread(s) per core:                   1
Core(s) per socket:                   28
Socket(s):                            2
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3488.5249
CPU min MHz:                          1500.0000
BogoMIPS:                             5490.03
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca debug_swap
L1d cache:                            1.8 MiB (56 instances)
L1i cache:                            1.8 MiB (56 instances)
L2 cache:                             28 MiB (56 instances)
L3 cache:                             128 MiB (8 instances)
NUMA node(s):                         8
NUMA node0 CPU(s):                    0-6
NUMA node1 CPU(s):                    7-13
NUMA node2 CPU(s):                    14-20
NUMA node3 CPU(s):                    21-27
NUMA node4 CPU(s):                    28-34
NUMA node5 CPU(s):                    35-41
NUMA node6 CPU(s):                    42-48
NUMA node7 CPU(s):                    49-55
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.2.6.post1+cu128torch2.7
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.3.14
[pip3] nvidia-cuda-cupti-cu12==12.8.57
[pip3] nvidia-cuda-nvrtc-cu12==12.8.61
[pip3] nvidia-cuda-runtime-cu12==12.8.57
[pip3] nvidia-cudnn-cu12==9.7.1.26
[pip3] nvidia-cufft-cu12==11.3.3.41
[pip3] nvidia-cufile-cu12==1.13.0.11
[pip3] nvidia-curand-cu12==10.3.9.55
[pip3] nvidia-cusolver-cu12==11.7.2.55
[pip3] nvidia-cusparse-cu12==12.5.7.53
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.8.61
[pip3] nvidia-nvtx-cu12==12.8.55
[pip3] onnxruntime==1.22.1
[pip3] pyzmq==27.0.0
[pip3] sentence-transformers==5.0.0
[pip3] torch==2.7.0+cu128
[pip3] torchaudio==2.7.0+cu128
[pip3] torchvision==0.22.0+cu128
[pip3] transformers==4.53.1
[pip3] triton==3.3.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
Neuron SDK Version           : N/A
vLLM Version                 : 0.9.2
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
        GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      49-55   7               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
OMP_NUM_THREADS=1
NCCL_CUMEM_ENABLE=0
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
CUDA_MODULE_LOADING=LAZY
```

**Comment by DarkLight1337** [comment]
You can set `bad_words` when passing sampling parameters

**Comment by vadimkantorov** [comment]
I also wonder, how to do LogitsProcessor to force max-model-tokens generation (e.g. forcing ignoring eos_id token until the very last token)? This is useful in VRAM stress-testing scenarios, to force maximal VRAM consumption. Or maybe is it possible to force this mode using SamplingParams?

**Comment by DarkLight1337** [comment]
cc @njhill

**Comment by njhill** [comment]
@Famondir custom logits processors aren't yet supported in V1 but work is in progress and support should be added very soon. See https://github.com/vllm-project/vllm/pull/19912. You will not be able to pass in a LP implementation in the sampling params. Instead you will specify the logits processor(s) at deployment time and they can read custom sampling parameters from requests. We will support backwards compatibility v0 logits processors via an adapter. cc @afeldman-nm.

> I also wonder, how to do LogitsProcessor to force max-model-tokens generation (e.g. forcing ignoring eos_id token until the very last token)? This is useful in VRAM stress-testing scenarios, to force maximal VRAM consumption. Or maybe is it possible to force this mode using SamplingParams?

@vadimkantorov this is already possible, you can either pass `ignore_eos=True` or set `min_tokens` to be equal to `max_tokens`. In the former case it may actually include the EOS token(s) in the output, in the latter case it will just not sample the EOS token until `min_tokens` have been generated.

**Comment by vadimkantorov** [comment]
> in the latter case it will just not sample the EOS token until `min_tokens` have been generated.

@njhill We've just tried that (min_tokens == max_tokens == 20580, all penalty params set to zero) with vllm 0.9.2, but the `len(request_output.outputs[0].completion.token_ids)` is still much smaller than `min_tokens`... Seems not working

OTOH setting `ignore_eos=True` worked

I do not fully understand how `min_tokens` relates to penalties and what exactly this code does:
https://github.com/vllm-project/vllm/blob/34a20c49b3f81f64133428b3a0d62309db1256f9/vllm/model_executor/layers/sampler.py#L345-L351

Is this code obsole for V1? The sampler in V1:
https://github.com/vllm-project/vllm/blob/34a20c49b3f81f64133428b3a0d62309db1256f9/vllm/v1/sample/sampler.py#L57
seems to have no mention of `min_tokens` penalty

Instead I found in V1 MinTokensLogitsProcessor:
https://github.com/vllm-project/vllm/blob/34a20c49b3f81f64133428b3a0d62309db1256f9/vllm/v1/sample/logits_processor.py#L504-L519

but I'm not sure if I need to enable it somehow...

@njhill From our test, it seems that in V1 `min_tokens` set to the same value as `max_tokens` by itself, seems broken / not doing anything

---

Meanwhile `min_tokens` does not function, for `ignore_eos=True`, is it possible to also somehow add EOS to the list of bad/stop tokens to prevent it from generation completely? As I'm worried that many astray EOS in the middle of tokens can affect some varlen packing downstream

**Comment by njhill** [comment]
Thanks @vadimkantorov this may be a recent regression, the min_tokens handling was recently moved to a logits processor as you found but nothing special should be needed to enable/use it.

I'll investigate and/or open an issue for the apparent not-working `min_tokens`. I'd be surprised if we don't have CI coverage for this, but if that's the case that also needs to be addressed asap.

cc @afeldman-nm

**Comment by njhill** [comment]
Opened an issue to track this: https://github.com/vllm-project/vllm/issues/21950
